### DISCUSSION TITLE
HTML content not uniform between PDF & DOCX + Unwanted page breaks in PDF

### DISCUSSION THREAD

**Discussion by prshnt0071** [eclipse-birt/birt#1130] [Q&A]
We created a separate report for PDF with Orientation as RTL and a separate report to generate the DOCX.

However, we are facing below issues when the report is extracted in PDF -
1. The same HTML content are not uniform in the DOCX as well as in PDF. Please see below -
![Untitled](https://user-images.githubusercontent.com/118506393/205066060-027071a8-9772-4d0c-a66a-6d435c4803eb.png)

2. Unexpected page break happening in the PDF output which does not occur in the DOCX
Note: The Layout = 'Auto' in our case
![Untitled1](https://user-images.githubusercontent.com/118506393/205068200-1515a7b7-1e50-42ce-8227-73fe04567020.png)

@All - Thanking you all in advance!! I would wait for your expertise inputs.

_Originally posted by @prshnt0071 in https://github.com/eclipse/birt/discussions/1114#discussioncomment-4283470_

**Comment by hvbtup** [comment]
I don't know if this is related to the page-break issue in any way, but usually you should use fixed layout instead of auto layout for all PDF reports and all reports which are designed for a fixed page size (not for a resizable browser window).

**Reply by prshnt0071** [reply]
@hvbtup Thanks!
I will ask my developer to try using fixed layout for PDF reports and check if we are able to kill Issue #2, but, any thoughts on Issue #1 related to HTML content being distorted in PDF?

**Reply by prshnt0071** [reply]
> I don't know if this is related to the page-break issue in any way, but usually you should use fixed layout instead of auto layout for all PDF reports and all reports which are designed for a fixed page size (not for a resizable browser window).

@hvbtup  We tried using the fixed layout for PDF report, but it seems there is something that we have not done correctly. We still observed page break as seen in the below snapshot.
![Untitled](https://user-images.githubusercontent.com/118506393/209347271-f445c2fd-029f-4b9c-a0f0-025d78c3ec8c.png)


Below I have attached the .rptdesign and the HTML content that is getting pushed to next page. Can you please review the page break rules applied by us in .rptdesign and guide us?

[Credit_test.zip](https://github.com/eclipse/birt/files/10295447/Credit_test.zip)

**Reply by prshnt0071** [reply]
@hvbtup @speckyspooky Any help on this?

**Comment by speckyspooky** [comment]
The isse with the HTML-table isn't so easy to say if you change this it will work.
The reason is that the Browser with HTML and DOCX is able to interpret your stored HTML-text/HTML-table because the internal engines understand your instructions.

The PDF-emitter isn't able to understand all CSS-instruction e.g CSS3 instructions are not usable.

If you get a sample report and sample HTML-table of your stored data I can take a look into it.
The investigation would be to figure out the responsible CSS-properties of your issue and check weather there are alternative attributes which the PDF-emitter can use.

But currently nobody can say based on your screens which changes are necessary.

**Reply by prshnt0071** [reply]
> The isse with the HTML-table isn't so easy to say if you change this it will work. The reason is that the Browser with HTML and DOCX is able to interpret your stored HTML-text/HTML-table because the internal engines understand your instructions.
> 
> The PDF-emitter isn't able to understand all CSS-instruction e.g CSS3 instructions are not usable.
> 
> If you get a sample report and sample HTML-table of your stored data I can take a look into it. The investigation would be to figure out the responsible CSS-properties of your issue and check weather there are alternative attributes which the PDF-emitter can use.
> 
> But currently nobody can say based on your screens which changes are necessary.

@speckyspooky The problem was caused in PDF due to a fix width that was present in the colspan and the only way for us to handle was to remove that dynamically while saving the HTML content in the database.
We were able to achieve the same and is getting displayed properly seen in the below screenshot 

![image](https://user-images.githubusercontent.com/118506393/209347865-ff0ba83e-175c-46df-b394-3c9ad670c2cb.png)

However, I am still curious to know if there is any other way to handle the above through BIRT directly without manipulating the HTML content. Below is the HTML content with width in it for your reference and the associated .rptdesign
[Credit_test.zip](https://github.com/eclipse/birt/files/10295493/Credit_test.zip)

[HTML Content_with colspan.zip](https://github.com/eclipse/birt/files/10295536/HTML.Content_with.colspan.zip)

**Reply by speckyspooky** [reply]
I spend time to verify your report and data.
(A short comment, it was not so easy to get yur report runable due to missing data connection and library usages.)

At frist, you will get the same results if you use the web browser preview because so both processing steps of Run-Task and Render-Task will be done (PDF = DOCX with page breaks) based on the same rptdocument of rendering process.
you start the output generation directly from BIRT-designer or eclipse therefore you got such different results.

In general of BIRT and page break handling BIRT tries to hold text with unknown length together with maximum size of this specific element (it has to do with the calculation of the element height and based on it the page break calculation).
In your case you put a **very large HTML-text-structure** with different information in an element and BIRT hold it together till the space is over. Therefore you get the large empty space.


**Hint to be used tables or grids**
- You use many tables without detail rows only to bind datasets -> in this case you can use directly a grid and bind the dataset there. 

**Hint to be further HTML issue (colspan & col with)**
- independent of your issue with the colspan (your solution is fine) will you get an additional one because your column width will not displayed correctly in PDF output.
![output_pdf_docx_column_issue](https://user-images.githubusercontent.com/41593722/209935777-e3d89fa7-2cd6-4d5e-b168-733405193335.png)


**Suggestion - layout refactoring**
- My suggestion to solve many of your problems will be to split the many many information of your HTML-structure to be more normalized data and display it in a normal BIRT-table because so you get full access of the structure handling.
- You won't be able to solve all issues with your big HTML-structure.

Step 1: I would split your HTML-data based of your example into 3 blocks
Step 2: I would split the 3rd block (=table) into standard BIRT-table including the dataset
![html_content_split](https://user-images.githubusercontent.com/41593722/209935722-989b80aa-51b6-45c1-996e-f34fcf1cadfc.png)

Yes, I know you have written that you use and HTML-editor. So I think you need a very short ETL-process to split the editor data into the normalized data and so you can save it into your database.

Here I add based on your report my test report with the grids instead of tables:
[Credit_Ar_pdf_copy.zip](https://github.com/eclipse/birt/files/10318803/Credit_Ar_pdf_copy.zip)
